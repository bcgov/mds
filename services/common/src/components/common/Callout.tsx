import React, { FC, ReactNode } from "react";

import { CALLOUT_SEVERITY } from "@mds/common/constants/strings";

interface CallOutProps {
  message: string | ReactNode;
  title?: string;
  severity: string;
}

const Callout: FC<CallOutProps> = ({ message, title, severity = CALLOUT_SEVERITY.info }) => {
  const formattedMessage = typeof message === "string" ? <p>{message}</p> : message;

  return (
    <div className={`bcgov-callout--${severity} bcgov-callout`}>
      {title && <p className="bcgov-callout-title">{title}</p>}
      {formattedMessage}
    </div>
  );
};

export default Callout;
