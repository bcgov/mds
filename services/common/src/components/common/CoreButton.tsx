import React, { FC } from "react";
import { Button, ButtonProps } from "antd";
import { ButtonType } from "antd/lib/button";

// DO NOT put antd button types in here: default, primary, ghost, dashed, link, text
declare const AdditionalTypes: ["tertiary", "filled-tertiary"];
// Additional doesn't exist at compile-time, I have not found a way around this. :(
// for now the two arrays should be duplicated.
const additionalTypes = ["tertiary", "filled-tertiary"];

type CoreCustomType = typeof AdditionalTypes[number];

// this one is the default because it's not used anywhere in the system,
// therefore it can be styled freely
const defaultButtonType = "dashed";

interface CoreButtonProps extends Omit<ButtonProps, "type"> {
  type: ButtonType | CoreCustomType;
}

const CoreButton: FC<CoreButtonProps> = ({ type, className, children, ...props }) => {
  const isAntdType = additionalTypes.includes(type);
  const buttonType = isAntdType ? defaultButtonType : (type as ButtonType);
  const buttonClassName = ["core-btn", `core-btn-${type}`, className].join(" ").trim();

  return (
    <Button {...props} type={buttonType} className={buttonClassName}>
      {children}
    </Button>
  );
};

export default CoreButton;
