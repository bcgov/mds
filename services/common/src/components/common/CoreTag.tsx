import React, { FC, ReactNode } from "react";
import { Row, Typography } from "antd";
import { Link } from "react-router-dom";

interface TagProps {
  text: string;
  icon: ReactNode;
  link?: string;
}

const CoreTag: FC<TagProps> = ({ text, icon, link }) => {
  const getText = () => {
    return link ? (
      <Link style={{ textDecoration: "none", color: "inherit" }} to={link}>
        {text}
      </Link>
    ) : (
      text
    );
  };

  return (
    <Row justify="space-between" align="middle" className="tag">
      {icon}
      <Typography.Text className="margin-medium--left">{getText()}</Typography.Text>
    </Row>
  );
};

export default CoreTag;
