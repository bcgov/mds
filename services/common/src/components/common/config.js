import RenderSelect from "../forms/RenderSelect";
import RenderDate from "./RenderDate";
import RenderYear from "./RenderYear";

// This file is anticipated to have multiple exports
// eslint-disable-next-line import/prefer-default-export
export const renderConfig = {
  SELECT: RenderSelect,
  DATE: RenderDate,
  YEAR: RenderYear,
};
