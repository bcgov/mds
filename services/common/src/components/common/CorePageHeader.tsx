import React, { FC, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Col, Row, Tabs, TabsProps, Typography } from "antd";
import { Link } from "react-router-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import CoreTag from "./CoreTag";

import CompanyIcon from "@mds/common/assets/icons/CompanyIcon";
import { faLocationDot } from "@fortawesome/pro-light-svg-icons";
import { getMineById } from "@mds/common/redux/selectors/mineSelectors";
import { fetchMineRecordById } from "@mds/common/redux/actionCreators/mineActionCreator";

interface BreadCrumb {
  route: string;
  text: string;
}

interface CorePageHeaderProps {
  entityLabel: string;
  entityType: string;
  mineGuid: string;
  current_permittee: string; // would be ideal to get this from the mine
  breadCrumbs?: BreadCrumb[];
  tabProps?: TabsProps;
}

const { Title, Text } = Typography;

const CorePageHeader: FC<CorePageHeaderProps> = ({
  mineGuid,
  current_permittee,
  entityLabel,
  entityType,
  breadCrumbs,
  tabProps,
}) => {
  const mine = useSelector((state) => getMineById(state, mineGuid));
  const dispatch = useDispatch();

  useEffect(() => {
    if (!mine) {
      dispatch(fetchMineRecordById(mineGuid));
    }
  }, [mineGuid]);

  return (
    <div className="core-page">
      <div className="view--header padding-lg--top padding-lg--sides core-page-header">
        <Row>
          <Col>
            {breadCrumbs.map((crumb) => {
              return (
                <React.Fragment key={crumb.route}>
                  <Link to={crumb.route} className="faded-text">
                    {crumb.text}
                  </Link>{" "}
                  /{" "}
                </React.Fragment>
              );
            })}
            <Text>
              {entityType} {entityLabel}
            </Text>
          </Col>
        </Row>
        <Row align="middle" gutter={16}>
          <Col>
            <Title level={1} className="margin-none">
              {entityType} {entityLabel}
            </Title>
          </Col>
          <Col>
            <CoreTag
              icon={<FontAwesomeIcon icon={faLocationDot} />}
              text={mine?.mine_name}
              link={GLOBAL_ROUTES?.MINE_DASHBOARD.dynamicRoute(mineGuid)}
            />
          </Col>
          <Col>
            <CoreTag icon={<CompanyIcon />} text={current_permittee} />
          </Col>
        </Row>
      </div>
      {tabProps && (
        <Tabs
          className="core-tabs fixed-tabs-tabs padding-md--top"
          {...tabProps}
          tabBarStyle={{ paddingLeft: 20, paddingRight: 20 }}
        />
      )}
    </div>
  );
};

export default CorePageHeader;
