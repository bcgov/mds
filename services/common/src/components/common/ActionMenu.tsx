import React, { FC, ReactNode } from "react";
import { Button, Dropdown, Modal } from "antd";
import CaretDownOutlined from "@ant-design/icons/CaretDownOutlined";
import DownOutlined from "@ant-design/icons/DownOutlined";
import { ITableAction } from "@mds/common/components/common/CoreTableCommonColumns";

export const deleteConfirmWrapper = (recordDescription: string, onOk: () => void) => {
  const title = `Confirm Deletion`;
  const content = `Are you sure you want to delete this ${recordDescription}?`;
  const modalContent = {
    title,
    content,
    onOk,
    okText: "Delete",
  };
  return Modal.confirm(modalContent);
};

export const generateActionMenuItems = (actionItems: ITableAction[], record) => {
  return actionItems.map((action) => {
    return {
      key: action.key,
      icon: action.icon,
      label: (
        <button
          type="button"
          disabled={action.disabled}
          className={`full actions-dropdown-button`}
          data-testid={`action-button-${action.key}`}
          onClick={(event) => action.clickFunction(event, record)}
        >
          <div>{action.label}</div>
        </button>
      ),
    };
  });
};

export interface IHeaderAction {
  key: string;
  label: string;
  icon?: ReactNode;
  clickFunction: () => void | Promise<void>;
}
// Looks like a button, intended for page-scope, not record-scope in the actions
export const ActionMenuButton: FC<{ buttonText?: string; actions: IHeaderAction[] }> = ({
  actions,
  buttonText = "Action",
}) => {
  const items = generateActionMenuItems((actions as unknown) as ITableAction[], null);

  return (
    <Dropdown menu={{ items }} placement="bottomLeft">
      <Button type="ghost" className="actions-dropdown-button">
        {buttonText}
        <DownOutlined />
      </Button>
    </Dropdown>
  );
};

interface ActionMenuProps {
  record: any;
  actionItems: ITableAction[];
  category: string;
}
const ActionMenu: FC<ActionMenuProps> = ({ record, actionItems, category }) => {
  const items = generateActionMenuItems(actionItems, record);
  return (
    <Dropdown menu={{ items }} placement="bottomLeft">
      <Button type="text" className="actions-dropdown-button">
        Actions
        <CaretDownOutlined alt={`${category} Actions`} />
      </Button>
    </Dropdown>
  );
};

export default ActionMenu;
