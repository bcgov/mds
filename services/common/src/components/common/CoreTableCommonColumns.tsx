import React, { ReactNode } from "react";
import Highlight from "react-highlighter";
import { dateSorter, formatDate, nullableStringSorter } from "@mds/common/redux/utils/helpers";
import { EMPTY_FIELD } from "@mds/common/constants/strings";
import { ColumnType } from "antd/lib/table";
import { Button, Dropdown } from "antd";
import { CaretDownOutlined } from "@ant-design/icons";
import { generateActionMenuItems } from "./ActionMenu";

export const renderTextColumn = (
  dataIndex: string,
  title: string,
  sortable = false,
  placeHolder = EMPTY_FIELD,
  width?: number
): ColumnType<any> => {
  return {
    title,
    dataIndex,
    key: dataIndex,
    render: (text: string) => <div title={title}>{text ?? placeHolder}</div>,
    ...(width !== undefined ? { width } : null),
    ...(sortable ? { sorter: nullableStringSorter(dataIndex) } : null),
  };
};

export const renderDateColumn = (
  dataIndex: string,
  title = "Date",
  sortable = false,
  format: (date: any) => string | null = null,
  placeHolder = EMPTY_FIELD
) => {
  const formatFunction = format ?? formatDate;
  return {
    title,
    dataIndex,
    key: dataIndex,
    render: (text) => <div title={title}>{formatFunction(text) || placeHolder}</div>,
    ...(sortable ? { sorter: dateSorter(dataIndex) } : null),
  };
};

export const renderCategoryColumn = (
  dataIndex: string,
  title: string,
  categoryMap: any,
  sortable = false,
  placeHolder = EMPTY_FIELD
) => {
  return {
    title,
    dataIndex,
    key: dataIndex,
    render: (text: string) => <div title={title}>{categoryMap[text] ?? placeHolder}</div>,
    ...(sortable ? { sorter: nullableStringSorter(dataIndex) } : null),
  };
};

export const renderHighlightedTextColumn = (
  dataIndex: string,
  title: string,
  regex: string,
  sortable = true
): ColumnType<any> => {
  return {
    title,
    dataIndex,
    key: dataIndex,
    render: (text: string) => {
      return <Highlight search={regex}>{text}</Highlight>;
    },
    ...(sortable ? { sorter: nullableStringSorter(dataIndex) } : null),
  };
};

export interface ITableAction {
  key: string;
  label: string;
  clickFunction: (event, record) => any;
  icon?: ReactNode;
}

export const renderActionsColumn = (
  actions: ITableAction[],
  recordActionsFilter?: (record, actions) => ITableAction[],
  isRowSelected = false,
  text = "Actions",
  dropdownAltText = "Menu"
) => {
  return {
    key: "actions",
    render: (record) => {
      const filteredActions = recordActionsFilter ? recordActionsFilter(record, actions) : actions;
      const items = generateActionMenuItems(filteredActions, record);

      return (
        <div>
          {items.length > 0 && (
            <Dropdown menu={{ items }} placement="bottomLeft" disabled={isRowSelected}>
              {/* // TODO: change button classname to something generic */}
              <Button data-cy="menu-actions-button" className="permit-table-button">
                {text}
                <CaretDownOutlined alt={dropdownAltText} />
              </Button>
            </Dropdown>
          )}
        </div>
      );
    },
  };
};
