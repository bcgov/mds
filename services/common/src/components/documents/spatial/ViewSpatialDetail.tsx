import React, { FC } from "react";
import { useSelector } from "react-redux";
import CoreTable from "../../common/CoreTable";
import { renderTextColumn } from "../../common/CoreTableCommonColumns";
import { formatDate } from "@mds/common/redux/utils/helpers";
import { getUserInfo } from "@mds/common/redux/selectors/authenticationSelectors";

export interface ViewSpatialDetailProps {
  spatialBundle: any[];
}
const ViewSpatialDetail: FC<ViewSpatialDetailProps> = ({ spatialBundle }) => {
  const { preferred_username } = useSelector(getUserInfo);
  return (
    <>
      <div> map goes here</div>
      <CoreTable
        size="small"
        rowKey="document_manager_guid"
        showHeader={false}
        dataSource={spatialBundle}
        columns={[
          renderTextColumn("document_name", ""),
          {
            dataIndex: "document_name",
            key: "fileType",
            render: (text) => {
              return text.split(".")[1].toUpperCase();
            },
          },
          {
            key: "uploadDate",
            render: () => {
              return formatDate(new Date());
            },
          },
          {
            key: "uploadUser",
            render: () => {
              return preferred_username;
            },
          },
        ]}
      />
    </>
  );
};

export default ViewSpatialDetail;
