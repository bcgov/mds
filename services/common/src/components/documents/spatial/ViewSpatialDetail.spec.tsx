import React from "react";
import { render } from "@testing-library/react";
import { PROJECTS } from "@mds/common/constants/reducerTypes";
import * as MOCK from "@mds/common/tests/mocks/dataMocks";
import { ReduxWrapper } from "@mds/common/tests/utils/ReduxWrapper";
import ViewSpatialDetail from "./ViewSpatialDetail";
import {
  spatialBundlesFromFiles,
  spatialDataReducerType,
} from "@mds/common/redux/slices/spatialDataSlice";
import { MAJOR_MINES_APPLICATION_DOCUMENT_TYPE_CODE } from "@mds/common/constants";

const spatialDocuments = MOCK.PROJECT_SUMMARY.documents.filter(
  (d) => d.project_summary_document_type_code === MAJOR_MINES_APPLICATION_DOCUMENT_TYPE_CODE.SPATIAL
);
const spatialBundles = spatialBundlesFromFiles(spatialDocuments);
const bundle_id = spatialBundles[0].bundle_id;

const initialState = {
  [PROJECTS]: {
    projectSummary: MOCK.PROJECT_SUMMARY,
  },
  [spatialDataReducerType]: {
    geoJsonData: MOCK.GEOJSON_FEATURE_DATA,
    bundle_id,
    spatialBundle: spatialBundles[0],
  },
};

// mock the map to not show the Suspense component during render/lazy loading
jest.mock("@mds/common/components/common/Map", () => {
  return jest.requireActual("@mds/common/components/common/LeafletMap");
});

describe("ViewSpatialDetail", () => {
  it("renders properly", () => {
    const { container } = render(
      <ReduxWrapper initialState={initialState}>
        <ViewSpatialDetail spatialDocuments={spatialBundles[0].bundleFiles} />
      </ReduxWrapper>
    );
    expect(container).toMatchSnapshot();
  });
});
