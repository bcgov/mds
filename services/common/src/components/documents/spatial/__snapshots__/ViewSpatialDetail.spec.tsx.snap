// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewSpatialDetail renders properly 1`] = `
<div>
  <div
    class="leaflet-container leaflet-grab leaflet-touch-drag"
    id="leaflet-map"
    style="height: 100%; width: 100%; z-index: 0; min-height: 200px; position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      >
        <div
          class="leaflet-layer "
          style="z-index: 1;"
        >
          <div
            class="leaflet-tile-container leaflet-zoom-animated"
            style="z-index: 18; left: 0px; top: 0px;"
          >
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://server.arcgisonline.com/ArcGIS/rest/services/World_Topo_Map/MapServer/tile/0/0/0"
              style="width: 256px; height: 256px; left: -36px; top: -81px;"
            />
          </div>
        </div>
      </div>
      <div
        class="leaflet-pane leaflet-shadow-pane"
      />
      <div
        class="leaflet-pane leaflet-overlay-pane"
      >
        <svg
          height="0"
          pointer-events="none"
          style="left: 0px; top: 0px;"
          viewBox="0 0 0 0"
          width="0"
        >
          <g>
            <path
              class="leaflet-interactive"
              d="M3 3L3 -3L-3 -3L-3 3z"
              fill="#3388ff"
              fill-opacity="0.2"
              fill-rule="evenodd"
              stroke="#3388ff"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-opacity="1"
              stroke-width="3"
            />
          </g>
        </svg>
      </div>
      <div
        class="leaflet-pane leaflet-marker-pane"
      />
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      >
        <div
          class="leaflet-control-zoom leaflet-bar leaflet-control"
        >
          <a
            aria-label="Zoom in"
            class="leaflet-control-zoom-in"
            href="#"
            role="button"
            title="Zoom in"
          >
            +
          </a>
          <a
            aria-label="Zoom out"
            class="leaflet-control-zoom-out leaflet-disabled"
            href="#"
            role="button"
            title="Zoom out"
          >
            âˆ’
          </a>
        </div>
      </div>
      <div
        class="leaflet-top leaflet-right"
      />
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      />
    </div>
  </div>
  <div
    class="ant-table-wrapper  core-table"
  >
    <div
      class="ant-spin-nested-loading"
    >
      <div
        class="ant-spin-container"
      >
        <div
          class="ant-table ant-table-small"
        >
          <div
            class="ant-table-container"
          >
            <div
              class="ant-table-content"
            >
              <table
                style="table-layout: auto;"
              >
                <colgroup />
                <tbody
                  class="ant-table-tbody"
                >
                  <tr
                    class="ant-table-row ant-table-row-level-0 fade-in"
                    data-row-key="5a5d71a5-8129-4678-bd2d-50ddb543d85a"
                  >
                    <td
                      class="ant-table-cell"
                    >
                      <div
                        class="document_name-column"
                        title=""
                      >
                        shape.dbf
                      </div>
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      DBF
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      Jun 26 2024
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      idir\\username
                    </td>
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0 fade-in"
                    data-row-key="942a1535-2e84-4d28-831c-5c8c510ef116"
                  >
                    <td
                      class="ant-table-cell"
                    >
                      <div
                        class="document_name-column"
                        title=""
                      >
                        shape.prj
                      </div>
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      PRJ
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      Jun 26 2024
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      idir\\username
                    </td>
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0 fade-in"
                    data-row-key="33013146-f3b1-4e44-83ef-dfc1d57b883a"
                  >
                    <td
                      class="ant-table-cell"
                    >
                      <div
                        class="document_name-column"
                        title=""
                      >
                        shape.shp
                      </div>
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      SHP
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      Jun 26 2024
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      idir\\username
                    </td>
                  </tr>
                  <tr
                    class="ant-table-row ant-table-row-level-0 fade-in"
                    data-row-key="1c77f818-3d7a-461c-902d-7546755f4661"
                  >
                    <td
                      class="ant-table-cell"
                    >
                      <div
                        class="document_name-column"
                        title=""
                      >
                        shape.shx
                      </div>
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      SHX
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      Jun 26 2024
                    </td>
                    <td
                      class="ant-table-cell"
                    >
                      idir\\username
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
